week05/clause_test.go

```diff
@@ -0,0 +1,82 @@
+package demo
+
+import (
+	"testing"
+)
+
+func emptyClauses(t testing.TB) *Clauses {
+	t.Helper()
+	return &Clauses{}
+}
+
+func nonEmptyClauses(t testing.TB) *Clauses {
+	t.Helper()
+	return &Clauses{
+		"a": 42.0,
+		"qwerty": "asdf",
+	}
+}
+
+func TestClauseString(t *testing.T) {
+	t.Parallel()
+
+	table := []struct {
+		name	string
+		cls 	*Clauses
+		exp   	string
+	}{
+		{name: "empty clauses", cls: emptyClauses(t), exp: ""},
+		{name: "non-empty clauses", cls: nonEmptyClauses(t) , exp: "\"a\" = %!q(float64=42) and \"qwerty\" = \"asdf\""},
+	}
+
+	for _, tt := range table {
+		t.Run(tt.name, func(t *testing.T) {
+			act := tt.cls.String()
+			exp := tt.exp 
+			if act != exp {
+				t.Fatalf("expected %v, got %v", exp, act)
+			}
+		})
+	}
+}
+
+func TestClauseMatch(t *testing.T) {
+	t.Parallel()
+
+	matchingModel := Model{
+		"a": 42.0,
+		"qwerty": "asdf",
+	}
+
+	nonMatchingModel := Model{
+		"a": []string{"This", "is", "a", "Go", "course"},
+		"qwerty": "qwerty",
+	}
+
+	table := []struct {
+		name	string
+		cls 	*Clauses
+		model	Model
+		exp   	bool
+	}{
+		// testing with an empty Clauses returns true, whatever you compare it with. 
+		// I would expect Match to only return true if both the model and clauses are empty,
```

Comment:
Think of this in terms of SQL.

If we had `Clauses{"id": 1}`, it would result in a SQL query that matches only users that have an `id` = `1`.

```sql
select * from users where id = 1
```

If we had an empty set of clauses, `Clauses{}`, then it would result in a SQL statement that had no `where` clauses and would match all users in the database.

```sql
select * from users
```